### PR TITLE
feat(frontend): output explain result as graphviz dot format

### PR DESCRIPTION
**Pull request by lyang24**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
add graphviz support for explain

**Please do not leave this empty!**
Add dot format output for explain, the output could be visualized as 
image

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
- Refer to a related PR or issue link (optional)
close https://github.com/risingwavelabs/risingwave/issues/18803

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

Release note
Public Preview: Do note that the structure and format of EXPLAIN FORMAT=DOT is considered unstable, and may change overtime.

After this PR, one new [explain option](https://docs.risingwave.com/docs/current/sql-explain/#explain-options) is supported
`FORMAT DOT`. it will output dot format for logical and physical plans.

examples:
```
    CREATE TABLE t (v1 int);
    explain (logical, format dot) SELECT approx_percentile(0.5) WITHIN GROUP (order by v1) from t;
```
result
```
    digraph {
        0 [ label = "LogicalAgg\laggs: [\"approx_percentile($expr1)\"]\l" ]
        1 [ label = "LogicalProject\lexprs: [\"t.v1::Float64 as $expr1\"]\l" ]
        2 [ label = "LogicalScan\ltable: \"t\"\lcolumns: [\"v1\"]\l" ]
        0 -> 1 [ ]
        1 -> 2 [ ]
    }
```

### CODE REVIEW

**Comment by kwannoel** [comment]
Thanks for your work in this PR.

The output spawns new graph nodes for properties of plan nodes. I think it can be improved, these properties should be a part of their plan nodes' graph node.

Perhaps it's hard to implement with the petgraph::Graph interface. You may do some further investigation to see if `petgraph::Graph` can support it, or other more suitable approaches. I'll take a look into it as well.

**Comment by lyang24** [comment]
> Thanks for your work in this PR.
> 
> The output spawns new graph nodes for properties of plan nodes. I think it can be improved, these properties should be a part of their plan nodes' graph node.
> 
> Perhaps it's hard to implement with the petgraph::Graph interface. You may do some further investigation to see if `petgraph::Graph` can support it, or other more suitable approaches. I'll take a look into it as well.

Thank you I have updated the pr. I had this approach to serialize every graph node as label, very simple but very inefficient. let me know if you found better options.

**Comment by kwannoel** [comment]
Hmm still looks a little off from the provided screenshot. Let me provide an example to make it clearer

Consider the following query plan:
```
    LogicalProject { exprs: [min(t.v1), sum($expr1)] }
    └─LogicalAgg { group_key: [$expr1], aggs: [min(t.v1), sum($expr1)] }
      └─LogicalProject { exprs: [(t.v1 + t.v2) as $expr1, t.v1] }
        └─LogicalScan { table: t, columns: [t.v1, t.v2, t._row_id] }
```

We'd expect the output to be something like this:
```
digraph G {
    1 [label="LogicalProject\nexprs: [min(t.v1), sum($expr1)]"];
    2 [label="LogicalAgg\ngroup_key: [$expr1]\naggs: [min(t.v1), sum($expr1)]"];
    3 [label="LogicalProject\nexprs: [(t.v1 + t.v2) as $expr1, t.v1]"];
    4 [label="LogicalScan\ntable: t\ncolumns: [t.v1, t.v2, t._row_id]"];

    1 -> 2;
    2 -> 3;
    3 -> 4;
}
```

Screenshot 2024-11-19 at 4 25 25 PM

**Comment by kwannoel** [comment]
Typically, our plan-nodes are structured with `childless_record`:
https://github.com/risingwavelabs/risingwave/blob/5b9cf248a15ce1af4381809d0a3bdc464c4d3d31/src/frontend/src/optimizer/plan_node/logical_now.rs#L59

That's just a `XmlNode`: https://github.com/risingwavelabs/risingwave/blob/5b9cf248a15ce1af4381809d0a3bdc464c4d3d31/src/frontend/src/optimizer/plan_node/utils.rs#L219

Next, if we look at the `explain` step: https://github.com/risingwavelabs/risingwave/blob/5b9cf248a15ce1af4381809d0a3bdc464c4d3d31/src/frontend/src/optimizer/plan_node/mod.rs#L660-L666
We can see that we construct a `Pretty::Record` for this node.

So we can just assign nodes to `Pretty::Record`, and avoid assigning it to other `Pretty` variants.
If we find `Pretty::Record` also contains some properties, rather than plan nodes, we can further add a regex (`Stream.+`, `Batch.+`, `Logical.+`) to make sure it's handled.

**Comment by kwannoel** [comment]
Feel free to let me know if you encounter any difficulties.

**Comment by lyang24** [comment]
> Typically, our plan-nodes are structured with `childless_record`:
> 
> https://github.com/risingwavelabs/risingwave/blob/5b9cf248a15ce1af4381809d0a3bdc464c4d3d31/src/frontend/src/optimizer/plan_node/logical_now.rs#L59
> 
> That's just a `XmlNode`:
> 
> https://github.com/risingwavelabs/risingwave/blob/5b9cf248a15ce1af4381809d0a3bdc464c4d3d31/src/frontend/src/optimizer/plan_node/utils.rs#L219
> 
> Next, if we look at the `explain` step:
> 
> https://github.com/risingwavelabs/risingwave/blob/5b9cf248a15ce1af4381809d0a3bdc464c4d3d31/src/frontend/src/optimizer/plan_node/mod.rs#L660-L666
> 
> 
> We can see that we construct a `Pretty::Record` for this node.
> So we can just assign graph nodes to `Pretty::Record`, and avoid assigning it to other `Pretty` variants.
> 
> Next, an `XmlNode` has the following structure:
> 
> ```
> #[derive(Clone)]
> pub struct XmlNode<'a> {
>     pub name: Str<'a>,
>     pub fields: CowAssocArr<'a>,
>     /// Currently, if fields have `XmlNode` with children,
>     /// they will not be considered during linearization.
>     pub(crate) fields_is_linear: bool,
>     pub children: Vec<Pretty<'a>>,
> }
> ```
> 
> We can append the `fields` to the Graph Node's label. `Children` will just have an edge from this Graph Node, and have their own `GraphNode`, if they are `Pretty::Record`.

Thank you for the pointers I learned a lot, the code is actually simpler and the results are much nicer as well.

File: src/frontend/src/optimizer/plan_node/mod.rs [hunk 1]
```diff
@@ -697,6 +703,55 @@ impl Explain for PlanRef {
...
+    pretty: &Pretty<'_>,
+    graph: &mut Graph<String, String>,
+    nodes: &mut HashMap<String, NodeIndex>,
+    parent_label: Option<&str>,
+) {
+    if let Pretty::Record(r) = pretty {
+        let mut label = String::new();
+        label.push_str(&r.name);
+        if !r.fields.is_empty() {
+            label.push_str(" fields: ");
```
Comment (lyang24) [inline src/frontend/src/optimizer/plan_node/mod.rs:729]: The alternative is to add each field to a new line, this is much closer to the desired pictures provided
```
        for (k, v) in r.fields.iter() {
            label.push_str("\n");
            label.push_str(&k);
            label.push_str(": ");
            label.push_str(&serde_json::to_string(&PrettySerde(v.clone())).expect("failed to serialize plan to dot"));
        }
```
However, the visualized output does not look great in terms of alignment
image

File: src/frontend/planner_test/tests/testdata/output/explain_dot_format.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# This file is automatically generated. See `src/frontend/planner_test/README.md` for more information.
+- name: test dot output format (logical)
+  sql: |
+    CREATE TABLE t (v1 int);
+    explain (logical, format dot) SELECT approx_percentile(0.5) WITHIN GROUP (order by v1) from t;
+  explain_output: |
+    digraph {
+        0 [ label = "LogicalAgg fields: {\l  \"aggs\": [\l    \"approx_percentile($expr1)\"\l  ]\l}" ]
+        1 [ label = "LogicalProject fields: {\l  \"exprs\": [\l    \"t.v1::Float64 as $expr1\"\l  ]\l}" ]
+        2 [ label = "LogicalScan fields: {\l  \"columns\": [\l    \"v1\"\l  ],\l  \"table\": \"t\"\l}" ]
```
Comment (kwannoel) [inline src/frontend/planner_test/tests/testdata/output/explain_dot_format.yaml:10]: You probably just need `\l` after the last curly brace to align it
```suggestion
        2 [ label = "LogicalScan fields: {\l  \"columns\": [\l    \"v1\"\l  ],\l  \"table\": \"t\"\l}\l" ]
```

File: src/frontend/src/optimizer/plan_node/mod.rs (hunk 1 above)
Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/mod.rs:729]: see: https://github.com/risingwavelabs/risingwave/pull/19446/files#r1849893022

**Review by kwannoel — COMMENTED** [review]
Could we add more test cases? Like in https://github.com/risingwavelabs/risingwave/pull/19430.

Some further improvement would be to test large labels (stream_key has 200 columns for instance), and improve its `DOT` output, for instance doing truncation.
 This can be tracked and worked on in a separate issue. We have a good starting point from this PR.

Rest LGTM. Thanks for your efforts!

**Comment by lyang24** [comment]
> Could we add more test cases? Like in #19430.
> 
> Some further improvement would be to test large labels (stream_key has 200 columns for instance), and improve its `DOT` output, for instance doing truncation. This can be tracked and worked on in a separate issue and PR. We have a good starting point from this PR.
> 
> Rest LGTM. Thanks for your efforts!

Yep added the very long test case but i realize the output is not readable after convert it into graph because the serialization of the (nested) children nodes.
so I added 1 more optimization to skip serialization of chidren on dot output because directed graph can captures that with edges.

**Comment by lyang24** [comment]
> In our `DOT` output, the `StreamHashJoin` nodes are treated as identical, since their labels are the same. However, they are actually distinct plan nodes.
> We can use plan node id to distinguish them.

Great catch, thanks for the commit. `git cherry-pick 5f29e9c` do not work out on my laptop I included the change manually in another commit. Thanks again.

**Review by kwannoel — APPROVED** [review]

**Comment by kwannoel** [comment]
LGTM, thanks!

**Comment by kwannoel** [comment]
If you're interested there's a further extension of this task: https://github.com/risingwavelabs/risingwave/issues/19542.

We wish to use it in our kernel dashboard, to visualize distributed stream job fragments.

**Pull request merged**